### DISCUSSION TITLE
How to pass variable to app layout when calling livewire component directly from route

### DISCUSSION THREAD

**Discussion by ralphmorris** [livewire/livewire#4769] [1. Help] [answered]
I'm using Jetstream and want to pass a $header variable to the app layout when calling a livewire component directly from a route. 

I can't find any way to do this without defining a normal route to a controller which generates more files.

```
Route::get('/services/{service}', App\Http\Livewire\ServiceEdit::class)->name('dashboard.service.edit');
```

Is there any way to do this?

**Comment by zepfietje** [comment, marked as answer]
Hey @ralphmorris!

You can [pass data to your layout](https://laravel-livewire.com/docs/2.x/rendering-components#custom-layout) using `layout()` or `layoutData()` from your Livewire component.

Not 100% sure if this fixes your issue, as you mentioned you're using Jetstream.
Could you perhaps go into any specifics?

**Reply by ralphmorris** [reply]
Thanks for this @zepfietje ! Worked a charm. I don't know why I didn't see this in the docs.

I can now easily pass page titles to the Jetstream header in layouts.app.blade.php with

```
->layout('layouts.app', [
    'header' => 'My page title'
]);
```

**Reply by zepfietje** [reply]
Great! If `layouts.app` is already your default layout, you can use just this:
```php
->layoutData([
    'header' => 'My page title',
]);
```

**Comment by pronode** [comment]
In my case, I made a tiny PageComponent class which **automatically pushes all component properties to the layout**. Take a look:
```php
<?php

namespace App\Utils;

use Illuminate\Support\Str;
use Livewire\Component;

class PageComponent extends Component {

    protected string $layout; // Default layout

    public function render() {

        $viewName = str_replace('App\\Livewire\\Pages\\', '', get_class($this));
        $viewName = Str::snake($viewName);
        $viewName = str_replace('_', '-', $viewName);
        $viewName = str_replace('\\', '.', $viewName);

        $viewName = "livewire.pages." . $viewName;
        
        $attributes = get_object_vars($this);

        // Remove parent class properties using ReflectionClass:
        $reflectionClass = new \ReflectionClass($this);
        $parentClass = $reflectionClass->getParentClass();
        $parentClassProperties = $parentClass->getProperties();
        foreach ($parentClassProperties as $parentClassProperty) {
            $propertyName = $parentClassProperty->getName();
            unset($attributes[$propertyName]);
        }

        return view($viewName)->layout($this->layout, $attributes);
    }
}
```
Now, when you create a full-page component **all properties are passed to the layout** you defined in $layout property :)
```php
<?php

namespace App\Livewire\Pages;

use App\Models\Store;
use App\Utils\PageComponent;

class StorePage extends PageComponent
{
    protected $layout = 'layouts.store';

    public Store $store;

    public function mount($storeSlug)
    {
        $this->store = Store::where('slug', $storeSlug)->firstOrFail();
    }
}
```
Just keep your full-page components in **Livewire/Pages** directory
